### DISCUSSION TITLE
Headers problem

### DISCUSSION THREAD

**Discussion by wezy01** [cantaloupe-project/cantaloupe#536] [Q&A] [answered]
Hello everybody,

I'm currently creating a test environment for IIIF.
For this I'm using Cantaloupe image server 5.0.3.

I'll be using the authentication API and the delegates.rb file
When I want to access context['request_headers'] it returns NULL , 
other ones, like 'client_ip', 'request_uri', 'identifier' do work.

Also, if I want to do something like 
`return {
        'status_code' => 401,
        'challenge' => 'Bearer charset="UTF-8"'
    }
`
I'm getting the following error
`500 Internal Server Error
Cannot invoke "edu.illinois.library.cantaloupe.image.Info.getMetadata()" because "info" is null`

What could I do to solver these errors?

Thanks in advance,
Wezy

**Comment by adolski** [comment, marked as answer]
Hi @wezy01,

I can't reproduce your first problem:

```ruby
def pre_authorize(options = {})
  puts context['request_headers']
  true
end
```
The request headers are then printed to the console upon a request for http://localhost:8182/iiif/3/image.jpg/info.json:

```
{"Accept"=>"*/*", "User-Agent"=>"curl/7.77.0", "Host"=>"localhost:8182"}
```

----

I can confirm that the second problem involving returning 401 status from the `pre_authorize()` delegate method is a bug. I just pushed a fix to the `release/5.0` branch for inclusion in 5.0.5.

**Reply by wezy01** [reply]
For the first problem,
if I add the following in my delegates.rb

`headers = context['request_headers']

logger.debug context['authorization']`

and then, in the browser I go directly to http://10.2.0.22:8182/iiif/3/0000__0000__0000__0000_0000_0000_0168.tif/info.json (internal test environment), I receive the following error: 
`500 Internal Server Error
javax.script.ScriptException: org.jruby.embed.InvokeFailedException: java.lang.NullPointerException: Cannot invoke "edu.illinois.library.cantaloupe.resource.RequestContextMap$ValueGetter.get()" because the return value of "java.util.Map.get(Object)" is null`

Could it be that context['request_headers'] is emtpy?
And if so, how to resolve that?

**Comment by adolski** [comment]
Try this instead:

```ruby
logger.debug context.dig('request_headers', 'authorization')
```

**Reply by wezy01** [reply]
Well, the error is gone but the response is null
`2021-11-24 07:28:11,698 DEBUG [qtp1642030774-215] e.i.l.c.d.DelegateProxy [Logger.java:48] null`
Does this mean that the header is empty?

**Reply by adolski** [reply]
The header names are passed through to the delegate script as they are received, so it might be `authorization` or it might be `Authorization`. It will also be nil if the client has not sent it.

**Reply by wezy01** [reply]
Then I guess it's not set, because 'Authorisation' doesn't work either.
Any idea as of why the client doesn't send them?
I'm testing in Mirador 3 and also with the individual file ( http://10.2.0.22:8182/iiif/3/0000__0000__0000__0000_0000_0000_0168.tif/info.json ), again testing locally.
In the browser I'm getting the normal response 
`{"@context":"http://iiif.io/api/image/3/context.json","id":"http://10.2.0.22:8182/iiif/3/0000__0000__0000__0000_0000_0000_0168.tif","type":"ImageService3","protocol":"http://iiif.io/api/image","profile":"level2","width":11130,"height":12269,"maxArea":136553970,"sizes":[{"width":87,"height":96},{"width":174,"height":192},{"width":348,"height":383},{"width":696,"height":767},{"width":1391,"height":1534},{"width":2783,"height":3067},{"width":5565,"height":6135},{"width":11130,"height":12269}],"tiles":[{"width":512,"height":512,"scaleFactors":[1,2,4,8,16,32,64,128]}],"extraQualities":["bitonal","color","gray"],"extraFormats":["tif","gif"],"extraFeatures":["baseUriRedirect","canonicalLinkHeader","cors","jsonldMediaType","mirroring","profileLinkHeader","regionByPct","regionByPx","regionSquare","rotationArbitrary","rotationBy90s","sizeByConfinedWh","sizeByH","sizeByPct","sizeByW","sizeByWh"]}`

When looking in developer tools I'm seeing
![image](https://user-images.githubusercontent.com/3432279/143264942-88aeae3f-0265-4d85-ad90-9b9d95673065.png)
